### ISSUE TITLE
LeaderElection - error while releasing lock makes integration tests fail

### ISSUE THREAD

**Issue by derlin** [quarkiverse/quarkus-operator-sdk#505]
I am using failsafe to run the integration test. My tests were working well before the 5.X update, but now the tests pass, but failsafe itself crashes with:

```text
[INFO] Tests run: 10, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 22.271 s - in com.sicpa.ptf.extdboperator.ExternalDatabaseReconcilerIT
[INFO] Running com.sicpa.ptf.extdboperator.database.oracle.OracleDbActionIT
[WARNING] Tests run: 2, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 0 s - in com.sicpa.ptf.extdboperator.database.oracle.OracleDbActionIT
[INFO] Running com.sicpa.ptf.extdboperator.database.postgres.PostgresDbActionIT
[WARNING] Tests run: 4, Failures: 0, Errors: 0, Skipped: 4, Time elapsed: 0 s - in com.sicpa.ptf.extdboperator.database.postgres.PostgresDbActionIT
2023-02-07 08:09:24,538 INFO  [io.qua.ope.run.AppEventListener] (main) Quarkus Java Operator SDK extension is shutting down.
2023-02-07 08:09:24,538 INFO  [io.jav.ope.Operator] (main) Operator SDK 4.2.4 is shutting down...
2023-02-07 08:09:24,556 ERROR [io.fab.kub.cli.ext.lea.LeaderElector] (main) Exception occurred while releasing lock 'LeaseLock: default - external-db-operator (1e8ebe06-49eb-4c84-92ee-ee8609e942a1)' [Error Occurred After Shutdown]: io.fabric8.kubernetes.client.extended.leaderelection.resourcelock.LockException: Unable to update LeaseLock
	at io.fabric8.kubernetes.client.extended.leaderelection.resourcelock.LeaseLock.update(LeaseLock.java:102)
	at io.fabric8.kubernetes.client.extended.leaderelection.LeaderElector.release(LeaderElector.java:139)
	at io.fabric8.kubernetes.client.extended.leaderelection.LeaderElector.stopLeading(LeaderElector.java:120)
	at io.fabric8.kubernetes.client.extended.leaderelection.LeaderElector.lambda$null$1(LeaderElector.java:94)
...
2023-02-07 08:09:24,557 INFO  [io.jav.ope.LeaderElectionManager] (main) Stopped leading for identity: 1e8ebe06-49eb-4c84-92ee-ee8609e942a1. Exiting.
[DEBUG] Closing the fork 1 after not saying Good Bye.
...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
...
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-failsafe-plugin:3.0.0-M8:verify (default) on project 
[ERROR] org.apache.maven.surefire.booter.SurefireBooterForkException: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
...
[ERROR] Error occurred in starting fork, check output in log
[ERROR] Process Exit Code: 1
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:714)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:311)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:268)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeProvider(AbstractSurefireMojo.java:1311)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:1144)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:910)
[ERROR] 	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2(MojoExecutor.java:370)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute(MojoExecutor.java:351)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:215)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:171)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:163)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
[ERROR] 	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
[ERROR] 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:294)
[ERROR] 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
[ERROR] 	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
[ERROR] 	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:960)
[ERROR] 	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:293)
[ERROR] 	at org.apache.maven.cli.MavenCli.main(MavenCli.java:196)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)
```

I guess it is somewhat related to https://github.com/quarkiverse/quarkus-operator-sdk/issues/450, but I cannot use the same solution (disable when in dev mode), since I am not running tests in dev mode. There is no way to turn off leader election either, except by removing the feature altogether.

For info, my IT tests are annotated with `@QuarkusTest`, I set the `quarkus.operator-sdk.start-operator=false` property for tests only and have the following code to manually start the operator:

```java
@QuarkusTest
class SomeTestIT {

  @Inject
  KubernetesClient client;

  @Inject
  Operator operator;

  private static boolean isOperatorStarted = false;

  @BeforeEach
  public void startOperator() throws ClassNotFoundException {
    if (!isOperatorStarted) {
      operator.start();
      isOperatorStarted = true;
    }
    // ...
  }
}
```

Versions:

* quarkus-sdk: 5.0.4
* quarkus: 2.15.3.Final
* failsafe / surefire: 3.0.0-M8

**Comment by derlin** [comment]
I think I found the problem: https://github.com/java-operator-sdk/java-operator-sdk/blob/fb9dc1aeef2a5b0c9faf69286d39f16409f332f7/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/LeaderElectionManager.java#L79 

LeaderElection calls `System.exit(1)` on `stopLeading`. What I don't explain, though, is why I didn't have the issue before.

**Comment by metacosm** [comment]
I could introduce a configuration to deactivate leader election during tests…

**Comment by derlin** [comment]
A property to disable leader election (that can be set to true in the test profile) would indeed solve this, and potentially other problems. I wouldn't try to detect automatically if we are in tests though.

**Comment by metacosm** [comment]
Hi @derlin, could you give #506 a try and let me know if that addresses your issue?

**Comment by derlin** [comment]
@metacosm It seems to work.

Now that we have this, can't we remove the automatic disable in dev mode? My problem is the following: in case I decide to work on the operator, but I am connected to a cluster that has this same operator running (but in the cluster), then I will have conflicts between the two and wouldn't even notice until something goes wrong. To me, leader election was a form of security against this kind of errors. Now that leader-election doesn't work in dev mode, I have no way of avoiding this problem.

And just to be sure, do we absolutely need this `System.exit` ? Cause now, either I can test leader election, but any `mvn verify` will end up with status 1, or I cannot have any tests regarding leader election ...

**Comment by metacosm** [comment]
I will convert the new configuration to an enum (I think that should work) to allow people to select when leader election is active. Regarding the `System.exit`, I'll let @csviri answer.

**Comment by csviri** [comment]
@derlin  yes that is needed, since there is a clear goal not to have a parallel reconciliations running on a different instance, that can make all kind of problems (in general)

See E2E tests for leader election in JOSDK:

https://github.com/java-operator-sdk/java-operator-sdk/blob/9abcb3794a1d77a57af5a1d06fdbdc9952cf1fcf/sample-operators/leader-election/src/test/java/io/javaoperatorsdk/operator/sample/LeaderElectionE2E.java

Basically it's done by E2E tests with minikube. But I guess could work locally just running new processes (JVMs).

**Comment by derlin** [comment]
@csviri My tests were just testing that the lease existed and was taken - to ensure no one made a change that made leader election disappear. The fact that leader election works when active should be covered by your tests on the library :) but the thing is, because of this System.exit I cannot do that in integration tests: failsafe would complain and never exit properly.

This leaves me with only one option: turn off leader election in tests, and pray no one will suddenly change something making the tests connect to a real cluster instead of the dev-services one.

**Comment by csviri** [comment]
What I mean is that if you want to also test your operator in those scnenarios regarding leader election. 

>  failsafe would complain and never exit properly. 

not sure if I understand this

> This leaves me with only one option: turn off leader election in tests, and pray no one will suddenly change something making 
   the tests connect to a real cluster instead of the dev-services one.

I guess that would be a problem regardless of leader election turned on or not :)

**Comment by derlin** [comment]
failsafe launches a JVM fork to run the tests, and wait for it to exit properly. The whole process is explained here: https://maven.apache.org/surefire/maven-failsafe-plugin/examples/shutdown.html but from what I understood (and tested), if a JVM fork stops by itself (versus is stopped by failsafe) it is interpreted as a failure:

> SurefireBooterForkException: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?

The fact that the operator stops with a `System.exit(1)` thus will always make failsafe fail.

Concerning:

> I guess that would be a problem regardless of leader election turned on or not :)

Not exactly: what I want to avoid is for the operator to connect to a cluster where a released version of the operator already runs. In this case, having leader election also in tests will avoid problems: the operator would hang because it cannot get the lease, the test would fail and the developer would understand his mistake. This is why I am not fond of turning the leader election off in tests. That being said, I could also check manually in a beforeAll hooks whether a lease exists or not, instead of relying on leader election. This is a possible workaround, not ideal but could work.

**Comment by csviri** [comment]
I see, what comes to my mind is that we could have for these testing purpose a `LeadStopHandler`, an interface that could be implemented, where we will give a standard implementation and you could have and empty? or customized one for tests. 

But I guess there it would be the same problem, what if someone accidentally deploys the controller with the that special stop handler. 

> Not exactly: what I want to avoid is for the operator to connect to a cluster where a released version of the operator already 
   runs. In this case, having leader election also in tests will avoid problems: the operator would hang because it cannot get the 
   lease, the test would fail and the developer would understand his mistake. This is why I am not fond of turning the leader 
   election off in tests. That being said, I could also check manually in a beforeAll hooks whether a lease exists or not, instead of 
   relying on leader election. This is a possible workaround, not ideal but could work.

Not sure if I would solve such checks on this level. Basically what I hear is that there needs to be leader election in the tests because if case it is run against a cluster where is an operator running already it can cause problems. So as you said, you might want to check the lock or if the custom resource definition, or a deployment with the operator is present on the cluster. I understand that this is handy, but isn't it misused to solve a different problem?

**Comment by csviri** [comment]
So in general `System.exit()` needs to be there for production. For tests it is either leader election turned off, or it's replaced with some custom code - to have a generic solution. (Not, sure maybe something else I can't see ?)

**Comment by derlin** [comment]
For now, I think the ability to turn on and off the leader election mechanism would be enough. The PR by metacosm looks good, I am just not fond of this automatic turn off in dev mode. I would personally rather have the property, and choose myself whether I enable it in dev or not (it is so easy with quarkus profiles).

I will for now turn off leader election for tests, and simply check the existence of a Lease. If there is one, it means it isn't running against the devservice container and I can fail with an error.

Do we have a way to check if the leader election bean exists, but has been turned off ? This would be amazing, because my tests on leader election were only here for that (I once forgot the `@Unremovable` annotation on the bean, and leader election didn't work). If I could just inject a leader manager or something, and ensure it is there but off, then I am happy :)

**Comment by metacosm** [comment]
> Do we have a way to check if the leader election bean exists, but has been turned off ? This would be amazing, because my tests on leader election were only here for that (I once forgot the `@Unremovable` annotation on the bean, and leader election didn't work). If I could just inject a leader manager or something, and ensure it is there but off, then I am happy :)

@derlin how do you envision this working?

**Comment by derlin** [comment]
After looking at the code, I don't see an easy way to implement this wothout adding an attribute to leader election that makes a difference between null and disabled. But anyway, this is just a nice to have, for now I would be happy with your changes + the removal of the automatic disable in dev mode :)

**Comment by metacosm** [comment]
I've pushed a new version of the code that allows people to select the profiles in which leader election is active, with it being active by default in the prod profile.

**Comment by derlin** [comment]
@metacosm I tested it locally, seems fine by me :) Waiting for the release!

**Comment by metacosm** [comment]
Closed by #506.
